<link rel='stylesheet' href='../assets/css/main.css'/>

[<< back to main index](../README.md) 

# Lab 4.1 : Spark Dataframes


### Overview
First look at Spark Dataframes

### Depends On 
None

### Run time
20-30 mins


In [1]:
# initialize Spark Session
import os
import sys
top_dir = os.path.abspath(os.path.join(os.getcwd(), "../"))
if top_dir not in sys.path:
    sys.path.append(top_dir)

from init_spark import init_spark
spark = init_spark()

Initializing Spark...
Spark found in :  /home/ubuntu/apps/spark
Spark config:
	 spark.app.name=TestApp
	spark.master=local[*]
	executor.memory=2g
	spark.sql.warehouse.dir=/tmp/tmpin8u__8g
	some_property=some_value
Spark UI running on port 4040



## STEP 1: Load Clickstream data

In [2]:
clickstreamDF = spark.read.json("../data/click-stream/clickstream.json")
print(clickstreamDF)

DataFrame[action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string]


**==> Monitor Spark shell UI on port 4040**  
You may see something like this:

**==> Q : Why is Spark not lazy loading the JASON files?**

<img src="../assets/images/5.1a.png" style="border: 5px solid grey; max-width:100%;" />




## STEP 2 : Inspecting The Dataframe

**==> Print the schema of data frame**     


In [3]:
# TODO: Print the schema
# Hint : `printSchema()`

clickstreamDF.printSchema()

root
 |-- action: string (nullable = true)
 |-- campaign: string (nullable = true)
 |-- cost: long (nullable = true)
 |-- domain: string (nullable = true)
 |-- ip: string (nullable = true)
 |-- session: string (nullable = true)
 |-- timestamp: long (nullable = true)
 |-- user: string (nullable = true)



**==> Print / Dump the data contained within dataframe**  
Your output may look like this:

```
+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|blocked|campaign_12|   5|     facebook.com|ip_3|session_96|1420070400864|user_5|
|clicked| campaign_3|  54|sf.craigslist.org|ip_9|session_61|1420070401728|user_8|
...

```

In [4]:
# TODO: Print/Dump the data
# Hint : show()
clickstreamDF.show()

+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|blocked|campaign_12|   5|     facebook.com|ip_3|session_96|1420070400864|user_5|
|clicked| campaign_3|  54|sf.craigslist.org|ip_9|session_61|1420070401728|user_8|
|blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|
|clicked| campaign_6|  15|comedycentral.com|ip_9|session_49|1420070403456|user_4|
|blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|
|blocked| campaign_4| 171|   funnyordie.com|ip_1|session_92|1420070405184|user_9|
|blocked|campaign_17|  20|       amazon.com|ip_4|session_13|1420070406048|user_1|
|blocked|campaign_20|  78|        zynga.com|ip_5|session_36|1420070406912|user_3|
|blocked|campaig

**==> Explore methods available in Dataframe**  
Here is the Dataframe API : 
[Scala](http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.DataFrame)  /
[Java](http://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/DataFrame.html) / 
[Python](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=dataframe#pyspark.sql.DataFrame)



## STEP 3 : Querying Dataframe

**==> Show only click logs where the cost > 100**

Sample output

```
    +-------+-----------+----+-----------------+----+----------+-------------+------+
    | action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
    +-------+-----------+----+-----------------+----+----------+-------------+------+
    |clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
    |blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|
    |blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|
```

In [5]:
clickstreamDF.filter("cost > 100").show()

+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|
|blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|
|blocked| campaign_4| 171|   funnyordie.com|ip_1|session_92|1420070405184|user_9|
|blocked|campaign_19| 147|      nytimes.com|ip_2|session_65|1420070407776|user_6|
|clicked|campaign_16| 141|    wikipedia.org|ip_1|session_57|1420070409504|user_1|
|blocked| campaign_9| 153|       google.com|ip_3|session_22|1420070413824|user_9|
|blocked| campaign_8| 140|comedycentral.com|ip_8| session_4|1420070414688|user_2|
| viewed|campaign_20| 133|       google.com|ip_9|session_69|1420070416416|user_7|
+-------+-------

In [6]:
#Same thing
clickstreamDF.filter(clickstreamDF['cost'] > 150).show()


+-------+----------+----+--------------+----+----------+-------------+------+
| action|  campaign|cost|        domain|  ip|   session|    timestamp|  user|
+-------+----------+----+--------------+----+----------+-------------+------+
|blocked|campaign_4| 171|funnyordie.com|ip_1|session_92|1420070405184|user_9|
|blocked|campaign_9| 153|    google.com|ip_3|session_22|1420070413824|user_9|
+-------+----------+----+--------------+----+----------+-------------+------+



**==> Show the logs where action = clicked**  

In [7]:
# TODO: show the logs where action = clicked
# Hint : clickstreamDF.filter("action == '???'")
clickstreamDF.filter ("action == 'clicked'").show()

+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|clicked| campaign_3|  54|sf.craigslist.org|ip_9|session_61|1420070401728|user_8|
|clicked| campaign_6|  15|comedycentral.com|ip_9|session_49|1420070403456|user_4|
|clicked|campaign_16| 141|    wikipedia.org|ip_1|session_57|1420070409504|user_1|
|clicked| campaign_5|  74|       sfgate.com|ip_6|session_27|1420070412096|user_3|
+-------+-----------+----+-----------------+----+----------+-------------+------+



In [8]:
## TODO Another approach
clickstreamDF.filter(clickstreamDF['action'] == 'viewed').show()

+------+-----------+----+-----------------+----+----------+-------------+------+
|action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+------+-----------+----+-----------------+----+----------+-------------+------+
|viewed| campaign_5|  50|    wikipedia.org|ip_5|session_10|1420070408640|user_6|
|viewed|campaign_19|   0|sf.craigslist.org|ip_9|session_19|1420070410368|user_7|
|viewed| campaign_1|  24|        yahoo.com|ip_4|session_60|1420070411232|user_8|
|viewed|campaign_20| 133|       google.com|ip_9|session_69|1420070416416|user_7|
+------+-----------+----+-----------------+----+----------+-------------+------+



## Step 5 : Count the number of visits from each domain

Hint : `g = clickstreamDF.groupBy("column_name")`  
* Inspect methods on `g`  (use tab completion)  
* Then do `count` on `g`  
* Then do a show  
* So `clickstreamDF.groupby().count.show`  

In [9]:
## TODO: 
g = clickstreamDF.groupBy("domain")

## TODO: do the count
print(g.count())

## And finally show
print(g.count().show())

DataFrame[domain: string, count: bigint]
+-----------------+-----+
|           domain|count|
+-----------------+-----+
|      nytimes.com|    1|
|      youtube.com|    2|
|        zynga.com|    1|
|       google.com|    2|
|        yahoo.com|    1|
|     facebook.com|    1|
|          cnn.com|    1|
|    wikipedia.org|    3|
|       sfgate.com|    1|
|       amazon.com|    2|
|   funnyordie.com|    1|
|sf.craigslist.org|    2|
|comedycentral.com|    2|
+-----------------+-----+

None


## STEP 6 : Joining Dataframes

Let's load another data set `domain info`  
The data is in   `/data/click-stream/domain-info.json`  
The data looks like this:

```json
    {"domain":"amazon.com","category":"SHOPPING"}
    {"domain":"bbc.co.uk","category":"NEWS"}
    {"domain":"facebook.com","category":"SOCIAL"}
    ...
```

**==> Load the dataframe**

In [10]:
domainsDF = spark.read.json("../data/click-stream/domain-info.json")
print(domainsDF)


DataFrame[category: string, domain: string]


In [11]:
## TODO display domains data
## Hint : show()
domainsDF.show()

+-----------+-----------------+
|   category|           domain|
+-----------+-----------------+
|   SHOPPING|       amazon.com|
|       NEWS|        bbc.co.uk|
|     SOCIAL|     facebook.com|
|      PHOTO|       flickr.com|
|       NEWS|      foxnews.com|
|     SEARCH|       google.com|
|       NEWS|          npr.org|
|       NEWS|      nytimes.com|
|CLASSIFIEDS|sf.craigslist.org|
|       NEWS|       sfgate.com|
|     SOCIAL|      twitter.com|
|       NEWS|     usatoday.com|
|  REFERENCE|    wikipedia.org|
|     SEARCH|        yahoo.com|
|      VIDEO|      youtube.com|
+-----------+-----------------+



**==> Join both dataframes**

sample output
```
    +-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
    | action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|   category|           domain|
    +-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
    |clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|      VIDEO|      youtube.com|
    |blocked| campaign_2|   7|      youtube.com|ip_2|session_93|1420070412960|user_1|      VIDEO|      youtube.com|
    |blocked|campaign_17|  20|       amazon.com|ip_4|session_13|1420070406048|user_1|   SHOPPING|       amazon.com|
```

In [12]:

joined = clickstreamDF.join(domainsDF,  clickstreamDF["domain"] == domainsDF["domain"])
    
#see the results
joined.show()

+-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|   category|           domain|
+-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|      VIDEO|      youtube.com|
|blocked|campaign_12|   5|     facebook.com|ip_3|session_96|1420070400864|user_5|     SOCIAL|     facebook.com|
|clicked| campaign_3|  54|sf.craigslist.org|ip_9|session_61|1420070401728|user_8|CLASSIFIEDS|sf.craigslist.org|
|blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|  REFERENCE|    wikipedia.org|
|blocked|campaign_17|  20|       amazon.com|ip_4|session_13|1420070406048|user_1|   SHOPPING|       amazon.com|
|blocked|campaign_19| 147|      nytimes.com|ip_2|session_65|1420070407776|user_6|       NEWS|      nytim

**==> Note some rows are missing.  Which ones?  Why?**

**==> Do an outer join**   
Inspect the output, might look like this

**==> Can you explain the null values?**

Output:

```
    +-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
    | action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|   category|           domain|
    +-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
    |blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|       null|             null|
    |   null|       null|null|             null|null|      null|         null|  null|     SOCIAL|      twitter.com|
    |clicked| campaign_6|  15|comedycentral.com|ip_9|session_49|1420070403456|user_4|       null|             null|
```

In [13]:

joinedOuter = clickstreamDF.join(domainsDF,  clickstreamDF["domain"] == domainsDF["domain"], "outer")
joinedOuter.show()

+-------+-----------+----+-------------+----+----------+-------------+------+---------+-------------+
| action|   campaign|cost|       domain|  ip|   session|    timestamp|  user| category|       domain|
+-------+-----------+----+-------------+----+----------+-------------+------+---------+-------------+
|blocked|campaign_19| 147|  nytimes.com|ip_2|session_65|1420070407776|user_6|     NEWS|  nytimes.com|
|clicked|campaign_19| 118|  youtube.com|ip_4|session_36|1420070400000|user_9|    VIDEO|  youtube.com|
|blocked| campaign_2|   7|  youtube.com|ip_2|session_93|1420070412960|user_1|    VIDEO|  youtube.com|
|blocked|campaign_20|  78|    zynga.com|ip_5|session_36|1420070406912|user_3|     null|         null|
|blocked| campaign_9| 153|   google.com|ip_3|session_22|1420070413824|user_9|   SEARCH|   google.com|
| viewed|campaign_20| 133|   google.com|ip_9|session_69|1420070416416|user_7|   SEARCH|   google.com|
|   null|       null|null|         null|null|      null|         null|  null|     

## Step 7: Understanding Query Execution
We will use **explain** keyword

In [14]:
clickstreamDF.filter("cost > 100").explain(extended=True)

== Parsed Logical Plan ==
'Filter ('cost > 100)
+- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json

== Analyzed Logical Plan ==
action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string
Filter (cost#9L > cast(100 as bigint))
+- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json

== Optimized Logical Plan ==
Filter (isnotnull(cost#9L) AND (cost#9L > 100))
+- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json

== Physical Plan ==
*(1) Project [action#7, campaign#8, cost#9L, domain#10, ip#11, session#12, timestamp#13L, user#14]
+- *(1) Filter (isnotnull(cost#9L) AND (cost#9L > 100))
   +- FileScan json [action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] Batched: false, DataFilters: [isnotnull(cost#9L), (cost#9L > 100)], Format: JSON, Location: InMemoryFileIndex[file:/data/

In [15]:
joined.explain(extended=True)

== Parsed Logical Plan ==
Join Inner, (domain#10 = domain#264)
:- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json
+- Relation[category#263,domain#264] json

== Analyzed Logical Plan ==
action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string, category: string, domain: string
Join Inner, (domain#10 = domain#264)
:- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json
+- Relation[category#263,domain#264] json

== Optimized Logical Plan ==
Join Inner, (domain#10 = domain#264)
:- Filter isnotnull(domain#10)
:  +- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json
+- Filter isnotnull(domain#264)
   +- Relation[category#263,domain#264] json

== Physical Plan ==
*(2) BroadcastHashJoin [domain#10], [domain#264], Inner, BuildRight
:- *(2) Project [action#7, campaign#8, cost#9L, domain#10, ip#11, session#12

In [16]:
joinedOuter.explain(extended=True)

== Parsed Logical Plan ==
Join FullOuter, (domain#10 = domain#264)
:- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json
+- Relation[category#263,domain#264] json

== Analyzed Logical Plan ==
action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string, category: string, domain: string
Join FullOuter, (domain#10 = domain#264)
:- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json
+- Relation[category#263,domain#264] json

== Optimized Logical Plan ==
Join FullOuter, (domain#10 = domain#264)
:- Relation[action#7,campaign#8,cost#9L,domain#10,ip#11,session#12,timestamp#13L,user#14] json
+- Relation[category#263,domain#264] json

== Physical Plan ==
SortMergeJoin [domain#10], [domain#264], FullOuter
:- *(1) Sort [domain#10 ASC NULLS FIRST], false, 0
:  +- Exchange hashpartitioning(domain#10, 200), true, [id=#243]
:     +- FileScan json [action#7,c